docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+- Provides a read API to view results, which takes as input (`pseudoRandomSeed`, `limit`). Sorts all keys (trigger, upkeepID) with the `seed` and provides first `limit` results. We do not do FIFO here as potentially out of sync old results will block new results sent by the node for consensus.
+- Provides an API to remove (trigger, upkeepID) from the store
+
+### Metadata Store
+
+This component stores the metadata to be sent within observations to the network. There are three categories of metadata
+
+- Conditional Upkeep Instructions
+- Log recovery Instructions
+- Latest block history
+
+Every item has a TTL, and upon expiry items are just removed without any action.
+
+The store provides an add / view / remove API for other components.
+
+### Samples Observer
+
+Processes samples of upkeeps to be checked, using the latest block number as a trigger. It does the following procedures:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **metadata store**
+with conditional Upkeep instructions
+- Errors and ineligible results are ignored
+
+### Conditional Observer
+
+Processes coordinated block + upkeep payload coming from plugin and does the following:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **result store**
+- Errors and ineligible results are ignored
+
+### Log Observer
+
+Processes latest logs for active upkeeps. It does the following procedures:
+
+- Called every second by Log Trigger Ticker with latest logs as full payloads.
+- Filter out logs with block number that is older than threshold
+    - These are expected to be picked up by the log recoverer 
+- Pre-processes to filter the logs already present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into result store
+- If runner gave an error, put back into log retry ticker, to be performed after a fixed timeout (can extend to exponential backoff)
+
+<aside>
+💡 Note: Duplicate logs can enter this flow, this component should be able to handle that. This component can maintain in-memory state for processed logs and can also use that state for exponential backoff
+</aside>
+
+### Retry Observer
+
+**TODO**
+
+### Recovery Observer
+
+This flow gets as input an upkeepID and a logIdentifier. It is a backup flow which verifies that a log was fully processed. Logs can be put here from the node itself (through Log Recoverer) or from other nodes (through Plugin)
+
+- Input is an upkeepID, logTxHash, logIndex (optional checkBlockNum, checkBlockHash when passed from plugin)
+- Checks whether log should be processed
+    - Whether its part of the filters for that upkeep and it is still present on chain (DB call to log poller)
+    - Whether it is older from the latest block within a threshold (To prevent malicious node requesting recovery on a new log)
```

Comment:
This is not currently part of the recovery observer. These protections were added to the check pipeline as input validation where the pipeline won't return a result that is outside it operating parameters.